mixed-hcv
=========
[![DOI](https://zenodo.org/badge/29322913.svg)](https://zenodo.org/badge/latestdoi/29322913)

**mixed-hcv** is a bioinformatic pipeline implemented in Python that is designed to detect and quantify the presence of multiple hepatitis C virus (HCV) genotypes in Illumina MiSeq reads derived from a sample of HCV RNA.

## Background
The primary inputs of the pipeline are the paired-end FASTQ files that correspond to a single demultiplexed sample as generated by the Illumina MiSeq platform and manufacturer software (MiSeqReporter).

**mixed-hcv** uses [bowtie2](http://bowtie-bio.sourceforge.net/bowtie2) to map the short reads contained in the FASTQ inputs to a large (n=571) set of HCV reference genome sequences.  This reference set was generated by querying the NCBI Genbank nucleotide database for all full-length HCV genome sequence records, merging these sequences with the a genotype reference set curated by the [Los Alamos National Laboratory (LANL) HCV Sequence Database](http://hcv.lanl.gov), and removing similar genome sequences on the basis of phylogenetic distance.  Each reference genome is annotated with the HCV genotype and subtype, and Genbank accession number.  

The purpose of incorporating a large number of HCV references, specifically multiple reference genomes for a given HCV genotype and subtype, is to improve the chances that a short read will be successfully mapped to a reference despite the extensive sequence divergence among HCV infections.  Although iterative mapping against an adaptive reference can significantly improve mapping efficiency, we have found that using this approach in the context of mixed HCV infections can cause the multiple adaptive references to 'collide' so that they draw from the same subset of reads corresponding to the predominant genotype in the sample.  

The default reference set (`gb-ref+hg38_v2`) also includes the [hg38 human genome assembly](https://genome.ucsc.edu/cgi-bin/hgGateway?db=hg38).  This is done to exclude reads derived from human RNA that would otherwise be spuriously mapped to an HCV reference genome.  Because we are performing short read mapping with very permissive settings, such spurious mappings can occur.

*bowtie2* is executed with `--local` alignment option that tolerates partial matches against a reference.  The [SAM formatted](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2723002/) results generated by *bowtie2* is redirected to standard output, where it is captured and parsed by the Python script.  This default mode of operation is used to conserve disk space, since the SAM file is never written to the file system.  However, the user has the option to write the SAM files by specifying a relative or absolute path to a destination folder with the `--cache` option on the command line.  

Each line of the SAM output reports the name of the reference sequence to which the corresponding read was mapped most successfully.  If the read could not be mapped to any of the references, then this field is labelled with an asterisk (`*`) and ignored by **mixed-hcv**.  We extract the HCV genotype from the reference name (rname) to which each read was mapped.  Thus, reads that were mapped to different reference genomes can contribute to the same genotype count.  By default, only the first read of a pair is used to generate genotype counts - second reads are ignored.  However, both reads can be combined for more detailed outputs by alternative versions of this pipeline.

Mapped reads can be filtered by several criteria:
1. low bowtie2 map quality
2. first reads whose mate (second read) mapped to a different HCV genotype
3. length of matched region (sum length of `M` tokens in the [CIGAR](https://samtools.github.io/hts-specs/SAMv1.pdf) string) exceeds cutoff
4. low bowtie2 alignment score
By default, **mixed-hcv** only uses criteria 2 and 3 (`min_match_len=100`).


## Outputs

The primary output of CSV reports the number of reads mapped to each HCV genotype for the sample.  Specifically:
* `runname` - MiSeq run name extracted from the name of the parent directory where the input FASTQ files are located
* `sample` - the sample name, extracted from the first field of the underscore-delimited FASTQ filename
* `snum` - the sample number (numerical order in the `SampleSheet.csv` manifest), extracted from the second field of the underscore-delimited FASTQ filename
* `subtype` - an HCV genotype based on standard HCV nomenclature.  This field is labeled subtype instead of genotype because an earlier version of the pipeline reported the read counts at a finer granularity of individual HCV subtypes.  If the entry is blank (`,,`) then the subsequent count and percentage entries represent the reads that failed to map to any reference genome.
* `count` - the absolute number of reads that mapped to a reference genome of the given HCV genotype
* `total` - the total number of reads in the sample; the denominator for calculating the percentage entry
* `perc` - the percent of reads in the sample that mapped to the given genotype.


## Alternate pipelines

**Full-Genome Hit Pipeline:**

Aligns reads against reference fasta and counts hits.  Outputs sam file including unaligned reads, secondary alignments, and chimeric alignments.  Any first-mate that aligns to HCV\* or hg38\* reference sequence counts as a hit.  Ignores low mapping quality, low alignment score (AS field) alignments.

Execute `run.sh` to run entire pipeline.

* `batch-mpi.py`:  Aligns reads, counts hits.  Caches sam, per-sample hit count CSV to hard-drive.  Will not overwrite existing cached files.  You must execute this with MPI.
  
* `collectcat.py`:  Collates all per-sample hit count CSV files into a single per-run hit count CSV.


**Target-Region Hit Pipeline:**

Aligns reads against reference fasta, counts hits against target-regions, and outputs sequences that hit target regions.  Outputs sam file excluding unaligned reads, and reads where both mates do not map to the same reference.  If the merged sequence from both mates covers the target region over the minimum width threshold, it is considered a hit.  Ignores secondary, chimeric, low mapping quality alignments.  Target regions are defined in /mixed-hcv/data/gb-ref2.coords by default, but can be overridden via commandline.  The target region coordinates CSV file specifies the 0-based target region coordinates with respect to the reference sequence.

Execute run_HCVDeli.sh to run entire pipeline.

* `HCVDeli.py`:  Aligns reads, merges mates, counts hits.  Caches sam, per-sample target-region sequence count CSV to hard-drive.  Will not overwrite existing cached files. You may execute this script with or without MPI.
  
* `collectcat.py`:  Collates all per-sample hit target-region sequence count CSV files into per-run sequence count CSV.

**Random primer project pipeline:**

1. `batch-mpi.py`: Map short reads in FASTQ files to `gb-ref+hg38_v2` reference set; run with `--cache` option to save SAM outputs to files
2. `sam2aln.py`: Converts cached SAM file to a CSV of nucleotide variants aligned to HCV reference genomes
3. `aln2aafreq.py`: Converts aligned nucleotide variants to amino acid counts within specified targets (NS3, NS5a, and NS5b)
4. `merge_by_ref_gene.py`: Collates amino acid counts from multiple reference coordinate systems to a single H77 coordinate system.
5. `coverage_map.R`: Generates PDF plot summarizing coverage of H77 NS3, NS5a and NS5b (amino acid coordinates)
6. `aln2coverage.py`: Converts aligned nucleotide variants to H77 genome-wide coverage data
7. `aln2coverage.R`: Generates PDF plot summarizing coverage of H77 nucleotide coordinates (genome-wide)


## Reference Fastas:

- `gb-ref+hg38_v2`: The HCV+Human Genome+Human Mitochondria reference fasta leads to fewest false hits, but at 3GB, it is too large to upload to github.  Stored in `/macdatafile/mixed-hcv/ref/gb-ref+hg38_v2.fa`

- `gb-ref`: Only HCV sequences from genbank.  Some of the sequences don't have accessions.  Also stored in `/macdatafile/mixed-hcv/ref/gb-ref.fa`.

- `gb-ref2`:  Only HCV sequences from genbank.  All sequences will have accessions.  Also stored in `/macdatafile/mixed-hcv/ref/gb-ref2.fa`.


  
